quaternions.scad

```diff
@@ -20,51 +20,82 @@
 // Internal
 function _Quat(a,s,w) = [a[0]*s, a[1]*s, a[2]*s, w];
 
+function _Qvec(q) = [q.x,q.y,q.z];
+
+function _Qreal(q) = q[3];
+
+function _Qset(v,r) = concat( v, r );
+
+// normalizes without checking
+function _Qnorm(q) = q/norm(q);
+
+
+// Function: Q_is_quat()
+// Usage:
+//   if(Q_is_quat(q)) a=0;
+// Description: Return true if q is a valid non-zero quaternion.
+// Arguments:
+//   q = object to check.
+function Q_is_quat(q) = is_vector(q,4) && ! approx(norm(q),0) ;
+
 
 // Function: Quat()
 // Usage:
 //   Quat(ax, ang);
-// Description: Create a new Quaternion from axis and angle of rotation.
+// Description: Create a normalized Quaternion from axis and angle of rotation.
 // Arguments:
 //   ax = Vector of axis of rotation.
 //   ang = Number of degrees to rotate around the axis counter-clockwise, when facing the origin.
-function Quat(ax=[0,0,1], ang=0) = _Quat(ax/norm(ax), sin(ang/2), cos(ang/2));
+function Quat(ax=[0,0,1], ang=0) = 
+  assert( is_vector(ax,3) && is_num(0*ang), "Invalid input")
+  let( n = norm(ax) )
+  approx(n,0) 
+  ? _Quat([0,0,0], sin(ang/2), cos(ang/2))
+  : _Quat(ax/n, sin(ang/2), cos(ang/2));

```

Comment:
Very minor formatting quibble for future reference... In most the library, the preferred multi-line formatting for ternary operators has the `?` at the end of the test expression, on the same line, and the `:` is usually at the end of the truth result expression.  The following are common formatting idioms for ternary ops, but are not enforced:

```openscad
a = test(x)? short_result_expr1 : short_result_expr2;

b = test(x)? somewhat_longer_result_expr1 :
    somewhat_longer_result_expr2;

c = test1(x)? short_result_expr1 : 
    test2(x)? short_result_expr2 : 
    test3(x)? short_result_expr3 : 
    default_result_expr;

d = test(x)? let(
    // complicated multi-line calculations here
) result1 : let(
    // complicated multi-line calculations here
) result2;

e = test(x)? (
    // more complicated subexpressions here, possibly with multiple let() blocks.
    result1
) : (
    // more complicated subexpressions here, possibly with multiple let() blocks.
    result2
);
```

Having said all that, I'll still accept this submitted code formatting.

Reply:
However, indentation really *should* be 4-spaces per level.

I used to use a tab for each indentation level and used an editor that showed tabs indented equivalent to 4 spaces. @adrianVmariano used 3 or 4 spaces per indent, depending on the bit of code. This got rather chaotic and problematic eventually. We agreed to standardize on 4-space indents as a result. I converted all the tabs to 4-spaces, and we've been slowly fixing inconsistent space indention as we edit since then.